apollo-client/src/main/java/com/ctrip/framework/apollo/internals/AbstractConfigRepository.java

```diff
@@ -73,4 +73,9 @@ protected void fireRepositoryChange(String namespace, Properties newProperties)
       }
     }
   }
+
+  @Override
+  public void initialize() {
+    this.sync();
```

Comment:
Is it better to call trySync?

Reply:
Change to `trySync` here will slower startup. Let me try to illustrate it.

---

### 1. everything from the rate limiter in RemoteConfigRepository

If  we do not wait for 2 consecutive calls to method `sync`, the rateLimter in RemoteConfigRepository will **block current thread 500ms** on the second call because code `m_loadConfigRateLimiter.tryAcquire(5, TimeUnit.SECONDS)`, that is why so slower when startup.

https://github.com/apolloconfig/apollo-java/blob/bcc40531f32e74d0c639b5e5d2ab2995fd00c3d6/apollo-client/src/main/java/com/ctrip/framework/apollo/internals/RemoteConfigRepository.java#L182-L188

```mermaid
sequenceDiagram
RemoteConfigRepository ->> RemoteConfigRepository: sync
RemoteConfigRepository ->> RemoteConfigRepository: loadApolloConfig
RemoteConfigRepository ->> m_loadConfigRateLimiter: tryAcquire(5, TimeUnit.SECONDS)
alt if Acquire success
	m_loadConfigRateLimiter -->> RemoteConfigRepository: return immediately
else Acquire fail
	m_loadConfigRateLimiter -->> RemoteConfigRepository: block current thread 500ms
end
```

### 2. how to solve it?

There are 2 solution to speed up.

solution 1: forbid 2 consecutive calls to method `sync`, 

solution 2: use another method on m_loadConfigRateLimiter let it won't **block current thread 500ms** .

---

### 3. Back to talk about method `initialize` in DefaultConfig.

According to the code

https://github.com/apolloconfig/apollo-java/blob/bcc40531f32e74d0c639b5e5d2ab2995fd00c3d6/apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java#L72-L78

m_configRepository.getConfig() will invoke method `sync`.


When we change to code

```java
  private void initialize() {
    try {
      m_configRepository.initialize();
      updateConfig(m_configRepository.getConfig(), m_configRepository.getSourceType());
    } catch (Throwable ex) {
      Tracer.logError(ex);
      logger.warn("Init Apollo Local Config failed - namespace: {}, reason: {}.",
          m_namespace, ExceptionUtil.getDetailMessage(ex));
    } finally {
```

`m_configRepository.initialize();` and `m_configRepository.getConfig()` alway invoke method `sync`,
i.e 2 consecutive calls to method `sync`, that will let startup slower.

But the different in `m_configRepository.initialize();`, if we use `sync` instead of `trySync`, 
the exception will throw quickly, so `updateConfig(m_configRepository.getConfig(), m_configRepository.getSourceType());` won't execute,

i.e choose solution 1: forbid 2 consecutive calls to method `sync` to speed up.

That's why use `sync` instead of `trySync` here.

Maybe there is better solution here?

Reply:
I agree with you. Let's keep it as it is.